File: kyu3/prime_streaming_PG-13.rb
class Primes
    @PRIME = [2]
    MAX = 16000000
    arr = [false] + [true] * (MAX-2)
    arr[0] = nil
    
    arr.each.with_index do |b, n|
      next unless b
      x = 2*n + 1
      @PRIME << x
      (x*x..MAX).step(x*2) { |i| arr[i/2] = nil }
    end
    arr = nil
    
    def self.stream
      @PRIME.each
    end
  end

File: kyu3/prime_streaming_PG-13.py
def primes():
    MAX = 16000000 - 1
    SQ = int(MAX**(1/2)+1)
    arr = [False] + [True]*MAX
    for i in range(3,SQ,2):
        if arr[i]:
            arr[i*i::2*i]=[False]*(((MAX-(i*i))//(2*i))+1)     
    return [2] + [i for i in range(3,MAX,2) if arr[i]]
#I don't need to generate numbers more than once in this kata
class Primes:
    li = primes()
    @staticmethod
    def stream():
        yield from Primes.li 